tools/src/tools.ml

```diff
@@ -1292,3 +1292,492 @@ module ExtractCodeblocks = struct
+                  exp with
+                  pexp_desc =
+                    Pexp_apply
+                      {
+                        funct;
+                        args =
+                          [
+                            lhs;
+                            ( Nolabel,
+                              Ast_helper.Exp.apply template_funct migrated_args
+                            );
+                          ];
+                        partial;
+                        transformed_jsx;
+                      };
+                }
+              | _ ->
+                (* TODO: More elaborate warnings etc *)
+                (* Invalid config. *)
+                exp)
+            | _ -> Ast_mapper.default_mapper.expr mapper exp);
+      }
+    in
+    mapper
+
+  let migrate ~entryPointFile ~outputMode =
+    let path =
+      match Filename.is_relative entryPointFile with
+      | true -> Unix.realpath entryPointFile
+      | false -> entryPointFile
+    in
+    let result =
+      if Filename.check_suffix path ".res" then
+        let parser =
+          Res_driver.parsing_engine.parse_implementation ~for_printer:true
+        in
+        let {Res_driver.parsetree; comments; source} = parser ~filename:path in
+        match Cmt.loadCmtInfosFromPath ~path with
+        | None ->
+          Error
+            (Printf.sprintf
+               "error: failed to run migration for %s because build artifacts \
+                could not be found. try to build the project"
+               path)
+        | Some {cmt_extra_info = {deprecated_used}} ->
+          let mapper = makeMapper deprecated_used in
+          let astMapped = mapper.structure mapper parsetree in
+          Ok
+            ( Res_printer.print_implementation
+                ~width:Res_printer.default_print_width astMapped ~comments,
+              source )
+      else if Filename.check_suffix path ".resi" then
+        let parser =
+          Res_driver.parsing_engine.parse_interface ~for_printer:true
+        in
+        let {Res_driver.parsetree = signature; comments; source} =
+          parser ~filename:path
+        in
+        let mapper = makeMapper [] in
+        let astMapped = mapper.signature mapper signature in
+        Ok
+          ( Res_printer.print_interface ~width:Res_printer.default_print_width
+              astMapped ~comments,
+            source )
+      else
+        Error
+          (Printf.sprintf
+             "File extension not supported. This command accepts .res, .resi \
+              files")
+    in
+    match result with
+    | Error e -> Error e
+    | Ok (contents, source) when contents <> source -> (
+      match outputMode with
+      | `Stdout -> Ok contents
+      | `File ->
+        let oc = open_out path in
+        Printf.fprintf oc "%s" contents;
+        close_out oc;
+        Ok (Filename.basename path ^ ": File migrated successfully"))
+    | Ok _ -> Ok (Filename.basename path ^ ": File did not need migration")
```

Comment:
This case also needs to handle `outputMode`, otherwise [`tools_tests/test.sh`](https://github.com/rescript-lang/rescript/blob/f0b6de967ea2e1e9cb56579f5d5788437ae4e147/tests/tools_tests/test.sh#L45-L50) will create a `.res` file with contents `<filepath>: File did not need migration`.


```suggestion
    | Ok (contents, _) -> (
      match outputMode with
      | `Stdout -> Ok contents
      | `File -> Ok (Filename.basename path ^ ": File did not need migration"))
```